### PR TITLE
Cirrus: More tests to verify cache_images

### PR DESCRIPTION
**Pull request by cevich**
Depends on #3078 and #3239

***CIRRUS: TEST IMAGES***

Signed-off-by: Chris Evich <cevich@redhat.com>

### CODE REVIEW

**Comment by cevich** [comment]
***note to me***: The new image-check script should fail until #3192 merges

**Comment by cevich** [comment]
***Note***: Need to resolve mount-options issue first before merging this.

@edsantiago mind taking a peek while we wait?  I'm not sure if I should unify the original check function you wrote, and the new one here (they're slightly different).  Also, I'm sure you'll spot some "better ideas" for some of my bash-chicken-scratch :smile:

File: contrib/cirrus/check_image.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+MIN_MEM_MB=2000
+read JUNK TOTAL USED MEM_FREE JUNK <<<$(free -tm | tail -1)
+test_okay $MEM_FREE -ge $MIN_MEM_MB 'Minimum memory'
+
+test_okay $(type -P podman &> /dev/null; echo $?) -ne "0" "expect podman command NOT found on path" || let "RET+=1"
+
+MIN_ZIP_VER='3.0'
+VER_RE='.+([[:digit:]]+\.[[:digit:]]+).+'
+ACTUAL_VER=$(zip --version 2>&1 | egrep -m 1 "Zip$VER_RE" | sed -r -e "s/$VER_RE/\\1/")
+test_okay "$MIN_ZIP_VER" == $(echo -e "$MIN_ZIP_VER\n$ACTUAL_VER" | sort -V | head -1) "zip version greater/equal to $MIN_ZIP_VER" || let "RET+=1"
```
Comment (edsantiago) [inline contrib/cirrus/check_image.sh]: I think you mean `tail`, not head

File: contrib/cirrus/lib.sh [hunk 2]
```diff
@@ -114,6 +112,48 @@ req_env_var() {
     done
 }
 
+test_okay() {
+    # test may be called with 2 or 3 arguments
```
Comment (edsantiago) [inline contrib/cirrus/lib.sh]: Actually 3 or 4

File: contrib/cirrus/lib.sh.t [hunk 3]
```diff
@@ -83,6 +83,46 @@ test_rev "FOO BAR" 9 'FATAL: test_rev() requires $BAR to be non-empty'
...
+test_test_okay "foo" "bar" "" "" 9 'FATAL: test_okay() requires $RHS to be non-empty'
+test_test_okay "foo" "bar" "baz" "" 9 'FATAL: test_okay() requires $DESC to be non-empty'
+test_test_okay "foo" "bar" "baz" "fail" 2 'not ok - fail: foo bar baz: bar: binary operator expected'
+test_test_okay "foo" "-gt" "baz" "FAIL" 2 'not ok - FAIL: foo -gt baz: foo: integer expression expected'
+test_test_okay "1" "-ne" "0" 'Basic Logic' 0 'ok - Basic Logic: 1 -ne 0'
+test_test_okay "some string" "==" "other string" 'Hmmm' 1 'not ok - Hmmm: some string == other string'
+
+# positive tests
+test_test_okay "1" "-ne" "0" 'logic' 0 'ok - logic: 1 -ne 0'
+test_test_okay "" "-r" "$0" 'Darth Vader'"'"'s Credit Card Number' 0 'ok - Darth Vader'"'"'s Credit Card Number: -r contrib/cirrus/lib.sh.t'
```
Comment (edsantiago) [inline contrib/cirrus/lib.sh.t]: This doesn't work if the test is invoked as `./contrib/etc`, or `./lib.sh.t` from the same directory, etc. (And, the test name is completely weird).

**Comment by cevich** [comment]
Thanks Ed, yeah I know all about being swamped.  

I'll see about simplifying test_okay (and read your other comments) later in the week.

**Comment by rhatdan** [comment]
@cevich Whats the latest on this PR?

**Comment by cevich** [comment]
Actually, @edsantiago would you mind taking over this PR?  I'm going to be out on vacation next week, and having this additional validation merged will come in handy should images need to be updated/changed.

The most super-important thing here is the disk-space check, that was a really difficult problem to solve[*] and if it ever sneaks back in, introduces a nasty, non-obvious testing flake.

[*]Actual fix is embedded in the base-image, however it can be overridden anytime (by static or dynamic cloud metadata).

**Comment by edsantiago** [comment]
@cevich I was only able to get to this today, Thursday of a busy week, and it's taken me a long time to wrap my head around it. I think some of the code (mountopts) is no longer necessary. And, well, I just can't find a way to accept `test_okay`. It's a huge amount of complexity that could, IMO, be better handled by using the builtin `test` and better-worded diagnostics. When you get back, would you consider minimizing this a little, just making it the bare essentials?

**Comment by cevich** [comment]
> It's a huge amount of complexity that could, IMO, be better handled by using the builtin test and better-worded diagnostics.

Yep I believe that, was hoping you might wield thy +12 sword of complexity-cleavage to achieve the same end-goal.

It's okay though, there are other tests that could/should be added anyway.  Like confirming `ooe.sh` is installed/executable, and that the google systemd services are enabled on boot:

```
google-accounts-daemon.service             enabled        
google-clock-skew-daemon.service           enabled        
google-instance-setup.service              enabled        
google-network-daemon.service              enabled        
google-shutdown-scripts.service            enabled        
google-startup-scripts.service             enabled        
```

**Comment by cevich** [comment]
@edsantiago there, PTAL, I think you'll like this better.

File: contrib/cirrus/check_image.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+for REQ_UNIT in google-accounts-daemon.service \
+                google-clock-skew-daemon.service \
+                google-instance-setup.service \
+                google-network-daemon.service \
+                google-shutdown-scripts.service \
+                google-startup-scripts.service
+do
+    item_test "required $REQ_UNIT enabled" \
+        "$(systemctl list-unit-files --no-legend $REQ_UNIT)" = "$REQ_UNIT enabled"
```
Comment (edsantiago) [inline contrib/cirrus/check_image.sh]: Missing `|| let RET<etc>`? (I can't test any of this on my system to verify if it works)

Comment (cevich) [reply contrib/cirrus/check_image.sh]: oh woops, thanks.

**Pull request merged**